### ISSUE TITLE
Enable the Bridge to shovel messages one-way

### ISSUE THREAD

**Issue by dvdstelt** [Particular/NServiceBus.MessagingBridge#118]
Customers have been asking about configuring the Bridge to support shoveling messages only one way. In other words, have endpoints on one side (one transport), but not the other.

Another example is to have only a single endpoint per logical Bridge.

- Allows more fine-grained config, scaling, etc.
- Allows hosting Bridge within an existing endpoint
- Allows Bridge to be set up in scenarios where messages are created on a different transport, but never return.

However, there's validation in the Bridge that demands configuration to have endpoints on both sides. This still enables Bridge-per-endpoint, but endpoints are configured in multiple Bridges and thus become competing consumers. Especially in the cloud, this introduces additional transactional costs.

It's possible that if we just remove validation, it just works, but this needs to be verified.

Solution spike:

- https://github.com/Particular/NServiceBus.MessagingBridge/pull/295

**Comment by dvdstelt** [comment]
Currently you can only set up the Bridge according to the top part of the image. But ideally it should be possible to host them as the bottom part.

![image](https://user-images.githubusercontent.com/1257410/189367686-8b8e599a-e603-4ea6-a649-634e1a494c75.png)

**Comment by andreasohlund** [comment]
Makes sense, I think we should make this a special ”single endpoint” mode that only allows a single endpoint to be configured so that users have to be explicit about this

**Comment by ramonsmits** [comment]
Fixed in https://github.com/Particular/NServiceBus.MessagingBridge/releases/tag/2.0.2

**Comment by ramonsmits** [comment]
We are reverting the one-way change and release 2.0.3 to restore the previous behavior

**Comment by markgould** [comment]
This would be useful to us. We use the community Router package to route one-way messages that come from SQL triggers/procedures to endpoints running on Azure ServiceBus.

**Comment by markgould** [comment]
Any other info I can provide on this or is there a potential timeline for this so we can migrate off of the Router?

**Comment by helenktsai** [comment]
@markgould This issue is eligible to be picked up for a future release, but we can't provide any timeline. We may reach out if questions arise when we start working on it, but there is nothing specific right now. By all means, share any additional info/comments you have that you think are important to consider in the meantime.

**Comment by dvdstelt** [comment]
> With the Router we add a Tenant header to the messages.

Did you do that in the Router or in the sending endpoint? The Bridge doesn't support any modifications to messages at all, although that's also something we're thinking about. If you'd like this in the Bridge, how would you know what to put in the tenant header?

**Comment by markgould** [comment]
> > With the Router we add a Tenant header to the messages.
> 
> Did you do that in the Router or in the sending endpoint? The Bridge doesn't support any modifications to messages at all, although that's also something we're thinking about. If you'd like this in the Bridge, how would you know what to put in the tenant header?

Currently we do it based on the database name - for example it will be Product_Tenant so we can pull it out based on naming convention, but it would be easy to add a header from the sending endpoint (which is done from SQL stored procedures)

``` 
 public class TenantHeaderRule : IRule<ForwardSendContext, ForwardSendContext>
    {
        public Task Invoke(ForwardSendContext context, Func<ForwardSendContext, Task> next)
        {
            if (!context.IncomingInterface.Contains("_"))
                return next(context);

            var tenantId = context.IncomingInterface.Substring(context.IncomingInterface.LastIndexOf("_", StringComparison.CurrentCulture) + 1);
            context.ForwardedHeaders.Add("TenantId", tenantId);
            return next(context);
        }
    }

**Comment by markgould** [comment]
Any updates on this feature request? For now we're still using the community NServiceBus.Router.

**Comment by mikeminutillo** [comment]
Hi @markgould,

There is nothing new to report at this stage. This feature is still eligible for inclusion in a future release, but we can't give a concrete timeline on when that will be.

**Comment by markgould** [comment]
Any updates on when this shovel mode may be supported?

**Comment by awright18** [comment]
@markgould,
At the moment, there's no update to share. This feature remains under consideration for a future release, though we don't have a specific timeline yet.

**Comment by jpalac** [comment]
The ability to shovel messages one-way will be available in release 4.0.1. Closing this issue.

**Comment by jpalac** [comment]
Release [4.0.1](https://github.com/Particular/NServiceBus.MessagingBridge/releases/tag/4.0.1) is available.
